c10/util/BorrowedTensorTraits.h

```diff
@@ -0,0 +1,69 @@
+#pragma once
+
+#include <c10/core/TensorImpl.h>
+#include <cstring>
+
+namespace c10 {
+// This traits class is very much logically part of the private
+// implementation of TensorBase. It pulls memcpy shenanigans that rely
+// on knowing how TensorBase (and intrusive_ptr) are represented and
+// work. However, the safety of this is guarded with tests that should
+// be running under ASAN.
+template <typename TensorType>
+struct BorrowedTraits {
+  using repr_type = TensorType;
+  using raw_impl_pointer_type = TensorImpl*;
+
+  static TensorType nullRepr() {
+    // intrusive_ptr(std::nullptr_t) sets the internal representation
+    // to NullType::singleton() instead. We want actual nullptr bit
+    // pattern. We will not dereference it or run the
+    // TensorImpl/intrusive_ptr destructor.
```

Comment:
Mixing two different null representations seems fairly dangerous. Why do we need `nullptr` here instead of `NullType::singleton()`?

Reply:
to be able to differentiate "unset borrowed" from "set to undefined tensor"

Reply:
this is needed in order to implement `operator bool`. Also, it's not dangerous -- the nullptr bit pattern is never to be used as a Tensor; it indicates the lack of a Tensor.

Reply:
> this is needed in order to implement `operator bool`

Let me rephrase. Why is it desirable to have optional-like semantics, instead of defaulting to undefined?

> Also, it's not dangerous -- the nullptr bit pattern is never to be used as a Tensor; 

That's what makes it dangerous. The `nullpt`-valued `Tensor` can escape if used incorrectly and the `intrusive_ptr` code that explicity checks for `NullType::singleton` instead of `nullptr` will interpret it as a valid object and try to `decref` it.

Reply:
> The nullptr-valued Tensor can escape if used incorrectly and the intrusive_ptr code that explicity checks for NullType::singleton instead of nullptr will interpret it as a valid object and try to decref it.

Yes, if people use this primitive incorrectly then they will get crashes.

> . Why is it desirable to have optional-like semantics, instead of defaulting to undefined?

TensorRef in functorch needs it, see https://github.com/pytorch/pytorch/pull/95419

Reply:
> TensorRef in functorch needs it, see https://github.com/pytorch/pytorch/pull/95419

Are you saying that `TensorRef` happens to make this distinction, or it's important to the functionality? i.e. if you instead wrote `using TensorRef = at::OptionalTensorRef` would the tests fail?